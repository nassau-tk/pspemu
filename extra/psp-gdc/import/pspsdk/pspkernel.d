module pspsdk.pspkernel;

/*
#include <pspuser.h>
#include <pspiofilemgr_kernel.h>
#include <psploadcore.h>
#include <pspstdio_kernel.h>
#include <pspsysreg.h>
#include <pspkdebug.h>
#include <pspintrman_kernel.h>
#include <pspmodulemgr_kernel.h>
*/

/**
 * Set the $pc register to a kernel memory address.
 *
 * When the PSP's kernel library stubs are called, they expect to be accessed
 * from the kernel's address space.  Use this function to set $pc to the kernel
 * address space, before calling a kernel library stub.
 */
/*
#define pspKernelSetKernelPC()  \
{     \
	__asm__ volatile (      \
	"la     $8, 1f\n\t"     \
	"lui    $9, 0x8000\n\t" \
	"or     $8, $9\n\t"     \
	"jr     $8\n\t"         \
	" nop\n\t"              \
	"1:\n\t"                \
	: : : "$8", "$9");      \
	sceKernelIcacheClearAll(); \
}
*/

/**
 * Set the $pc register to a user memory address.
 *
 */
/*
#define pspKernelSetUserPC()  \
{     \
	__asm__ volatile (      \
	"la     $8, 1f\n\t"     \
	"li     $9, 0x7FFFFFFF\n\t" \
	"and    $8, $9\n\t"     \
	"jr     $8\n\t"         \
	" nop\n\t"              \
	"1:\n\t"                \
	: : : "$8", "$9");      \
	sceKernelIcacheClearAll(); \
}
*/