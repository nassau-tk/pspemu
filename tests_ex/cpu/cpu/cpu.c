#include <pspkernel.h>

PSP_MODULE_INFO("cpu test", 0, 1, 1);
PSP_MAIN_THREAD_ATTR(THREAD_ATTR_USER);

#define OP1(TYPE) int __attribute__((noinline)) op_##TYPE(int x       ) { int result; asm volatile(#TYPE " %0, %1" : "=r"(result) : "r"(x)); return result; }
#define OP2(TYPE) int __attribute__((noinline)) op_##TYPE(int x, int y) { int result; asm volatile(#TYPE " %0, %1, %2" : "=r"(result) : "r"(x), "r"(y)); return result; }

#define OPX_TEST_START() Kprintf("--\n");
#define OP1_TEST_X(TYPE, a) Kprintf("%s 0x%08X -> %d\n", #TYPE, a, op_##TYPE(a));
#define OP1_TEST(TYPE, a) Kprintf("%s %d -> %d\n", #TYPE, a, op_##TYPE(a));
#define OP2_TEST(TYPE, a, b) Kprintf("%s %d, %d -> %d\n", #TYPE, a, b, op_##TYPE(a, b));

#define OP2_TEST_SET(TYPE) \
	OPX_TEST_START(); \
	OP2_TEST(TYPE, 1, 7); \
	OP2_TEST(TYPE, 3, 1); \
	OP2_TEST(TYPE, -1, 0); \
	OP2_TEST(TYPE, -1, -10); \
	OP2_TEST(TYPE, 0, 0); \
	

OP2(max)
OP2(min)

OP2(add)
OP2(addu)
OP2(sub)
OP2(subu)

OP1(clo)
OP1(clz)

int main(int argc, char *argv[]) {
	OP2_TEST_SET(max);
	OP2_TEST_SET(min);
	OP2_TEST_SET(add);
	OP2_TEST_SET(addu);
	OP2_TEST_SET(sub);
	OP2_TEST_SET(subu);
	
	OPX_TEST_START();
	OP1_TEST_X(clo, 0x00000000)
	OP1_TEST_X(clo, 0x80000000)
	OP1_TEST_X(clo, 0xF0000000)
	OP1_TEST_X(clo, 0xF000000F)
	OP1_TEST_X(clo, 0xFFFF000F)
	OP1_TEST_X(clo, 0xF7FF000F)
	OP1_TEST_X(clo, 0xFFFFFFFF)

	OPX_TEST_START();
	OP1_TEST_X(clz, 0x00000000)
	OP1_TEST_X(clz, 0x0000000F)
	OP1_TEST_X(clz, 0xF000000F)
	OP1_TEST_X(clz, 0x0000007F)
	OP1_TEST_X(clz, 0xFFFFFFFF)

	return 0;
}