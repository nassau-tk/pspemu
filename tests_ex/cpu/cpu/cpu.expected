--
max 1, 7 -> 7
max 3, 1 -> 3
max -1, 0 -> 0
max -1, -10 -> -1
max 0, 0 -> 0
--
min 1, 7 -> 1
min 3, 1 -> 1
min -1, 0 -> -1
min -1, -10 -> -10
min 0, 0 -> 0
--
add 1, 7 -> 8
add 3, 1 -> 4
add -1, 0 -> -1
add -1, -10 -> -11
add 0, 0 -> 0
--
addu 1, 7 -> 8
addu 3, 1 -> 4
addu -1, 0 -> -1
addu -1, -10 -> -11
addu 0, 0 -> 0
--
sub 1, 7 -> -6
sub 3, 1 -> 2
sub -1, 0 -> -1
sub -1, -10 -> 9
sub 0, 0 -> 0
--
subu 1, 7 -> -6
subu 3, 1 -> 2
subu -1, 0 -> -1
subu -1, -10 -> 9
subu 0, 0 -> 0
--
clo 0x00000000 -> 0
clo 0x80000000 -> 1
clo 0xF0000000 -> 4
clo 0xF000000F -> 4
clo 0xFFFF000F -> 16
clo 0xF7FF000F -> 4
clo 0xFFFFFFFF -> 32
--
clz 0x00000000 -> 32
clz 0x0000000F -> 28
clz 0xF000000F -> 0
clz 0x0000007F -> 25
clz 0xFFFFFFFF -> 0